doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+  repeated string label = 7;
+  message Columns {
+  // The COLUMN clause will be split into `column`. For example, in the COLUMN
+  // clause "COLUMN NUMERIC(sepal_length), INDICATOR(ID)" will be saved in
+  // `columns` as ["NUMERIC(sepal_length)", "INDICATOR(ID)"]
+    repeated string columns = 1;
+  }
+  // `columns` contain a map of string to `Columns` For multiple COLUMN clauses
+  // like "COLUMN ... FOR deep_feature, COLUMN ... FOR wide_feature". They will
+  // be parsed as {"deep_feature": [...], "wide_feature": [...]}. For single
+  // column clause like "column ...", "feature_columns" will be used as the
+  // default map key.
+  optional map<string, Columns> columns = 8;
+  // `model_save` specifies the saved model path in the INTO/USING clause.
+  optional bool model_save = 9;
+  enum Type {
+    RUN = 0;
+    TRAIN = 1;
+    PREDICT = 2;
+    EXPLAIN = 3;
+    EVALUATE = 4;
+    SOLVE = 5;
+  }
+  optional Type type = 10;
+  // `predict_target` specifies the column to store predict result.
+  // For example: "iris.predict.class"
+  optional string predict_target = 11;
+  // `original_sql` is the original statement from that this `Statement` is
+  // generated. This can be for diagnostic purpose.
+  optional string original_sql string = 100;
+}
+
+message Program {
+  // `datasource` is the connection string to the database
+  optional string datasource = 1;
+  // `statements` is a list of `Statement` to be executed as a workflow
+  repeated Statement statements = 2;
+}
+```
+
+###### Golang
+
+The golang package `ir` should fill the `Program` message according to the parsed result.
+
+#### Workflow And Platform
+
+###### Python
+
+As decribed earlier, we would reimplement `workflow` in python. 
+
+The module `execute.py` should define at least the following functions:
+
+```python
+def execute(program:sqlflow_ir_pb2.Program) -> str:
+    """
+    `execute` executes the SQLFlow `program` and returns the job id
+    """
+    pass
+  
+def fetch(workflow_id:str):
+    """
+    `fetch` gets output of the workflow of `workflow_id` in an incremental way
+    """
+    pass
+  
+def step(statement:sqlflow_ir_pb2.Statement):
+    """
+    `step` dispatches the `statement` to the `sqlflow_submitter` package
+           according to `statement.type` and the platform configured
+    """
+    pass
+  
+def main(mode:str):
+    """
+    `main` parses a `sqlflow_ir_pb2.Program` from stdin and call `execute`
+           or `step` to execute the Program. 
+           Or get a string from stdin and call `fetch` to get the result.
+    """
+    pass
+  
+if __name__ == '__main__':
+    main(sys.argv[1] if len(sys.argv) > 1 else "execute")
+```
+
+The python function `execute` executes the SQLFlow `program` and should do at least the following tasks:
+
+- Do semantic checking for the `program` using `contracts.py`. **Diagnostics are required**.
+- Call `features.py:generate_analysis_steps` to generate additional data transforming steps. 
```

Comment:
For `TO TRAIN` statement, we always have an analysis step. Inside this step, we will generate the analysis SQL and execute.
Maybe we won't generate the analysis SQL before generating the workflow.